Haskell Parser Combinators/article.md

```diff
@@ -0,0 +1,1332 @@
+  | SBool    Bool  -- false, true
+  | SId      Identifier  -- foo
+  deriving (Show)
+```
+
+We also need an alias for the type which `megaparsec` uses.
+
+```hs
+type Parser = Parsec
+  Void  -- The type for custom error messages. We have none, so use `Void`.
+  String  -- The input stream type. Let's use `String` for now, but for better performance, you might want to use `Text` or `ByteString`.
+```
+
+### Booleans
+
+Let's start with the simplest of the parsers: `SBool`. We use `<|>` which tries one parser, and failing that, will try the other one.
+
+```hs
+bool :: Parser Bool
+bool = False <$ string "false" <|> True <$ string "true"
+```
+
+We will also make a parser to represent our `SExp`-specific variants. As we progress, we will add more parsers to this list.
+
+```hs
+atom :: Parser SExp
+atom = choice
+  [ SBool <$> bool
+  ]
+```
+
+Here we use `choice`, which is the same as running `<|>` between each element of the list. `choice [a, b, c] = a <|> b <|> c`.
+
+It's important to notice that `choice`, as well as many other useful combinators, are part of the [`parser-combinators`](https://hackage.haskell.org/package/parser-combinators-1.3.0) package, which Megaparsec conveniently re-exports for us. Should you fail to find a specific function in Megaparsec's documentation, chances are that it's defined in `parser-combinators` instead.
+
+Now, fire up GHCi and test that we can parse it! `megaparsec` provides the `parseTest` function which allows us to conveniently test our parsers.
+
+```hs
+>>> parseTest atom "false"
+SBool False
+
+>>> parseTest atom "hey"
+1:1:
+  |
+1 | hey
+  | ^^^
+unexpected "hey"
+expecting "false" or "true"
+```
+
+### Integers
+
+To parse integers, one strategy is to parse as many digits (chars whose value is between `'0'` and `'9'`) as possible. This will give us a list of characters, which we can then `read`,
+
+For this parser, we use the `some` function, which tries to run some given parser **1 or more times**. `numberChar` is exported from `Text.Megaparsec.Char` and parses any character that matches a digit. In regex, this is equivalent to `[0-9]+`. This will return us a `[Char]` (aka `String`), which we can `read` to make it into an `Integer`.
+
+```hs
+integer :: Parser Integer
+integer = read <$> some numberChar
+```
+
+We now change our `atom` to also include integers.
+
+```hs
+atom :: Parser SExp
+atom = choice
+  [ SBool <$> bool
+  , SInteger <$> integer
+  ]
+```
+
+Let's also test it in GHCi:
+
+```hs
+>>> parseTest atom "42"
+SInteger 42
+
+>>> parseTest atom "hey"
+1:1:
+  |
+1 | hey
+  | ^^^
+unexpected "hey"
+expecting "false", "true", or numeric character
+```
+
+Although the error message is good, wouldn't it better if it said "integer" instead of "numeric character"? For this, we can use **labels**, which allows us to give a name to a parser. To do this, we can simply change our parser like this:
+
+```hs
+integer :: Parser Integer
+integer = label "integer" $ read <$> some numberChar
```

Comment:
... or a `<?>` operator with priority of `0`

Reply:
Right, this also exists. I added a point about it.